kv/txn_coord_sender.go

```diff
@@ -600,7 +600,25 @@ func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.
 	// so grab a copy before unlocking.
 	txn := txnMeta.txn
 	tc.Unlock()
+
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = tc.clock.Now()
+	ba.CmdID = ba.GetOrCreateCmdID(ba.Timestamp.WallTime)
+	ba.Txn = txn.Clone()
+
 	if !proceed {
+		// Actively abort the transaction and its intents since we assume it's abandoned
+		ab := &roachpb.EndTransactionRequest{
+			Commit:  false,
+			Intents: txnMeta.intents(),
```

Comment:
You should also consider not using the wrapped sender for `EndTransaction`, but instead sending it through `TxnCoordSender` itself. That would save you a lot of special casing.

Something like

`tc.stopper.RunAsyncTask(func() { tc.Send(....) })`.